### PR TITLE
[Feat][Java] Initialize JAVA SDK: add INFO implementation

### PR DESCRIPTION
**Pull request by Thespica**
## Proposed changes

This PR aims to implement grpahinfo in Java with [fastFFI](https://github.com/alibaba/fastFFI).

## Known problems
- [x] Can't implement function which return `StdMap`
- [x] Don't konw how to get generic factory like `StdVctor<Property>`
- [x] Test is too simple.

### Before 7.30
Make graphinfo's test better

### 7.31-8.13
- CI
- Readers and test.

### 8.14-8.28
Writers and test.

### After 8.28
- Generate JavaDoc
- README in java
- Examples

## Related issue
#72

### CODE REVIEW

**Comment by Thespica** [comment]
`StdMap` error log:
```
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ gar-java ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 23 source files to /home/john/my-reop-gar/java/target/classes
com.alibaba.fastffi.annotation.FFICodeGenerationException: Cannot find a TypeDef for 
        std::map<std::string, GraphArchive::VertexInfo> -> (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.a
libaba.graphar.stdcxx.StdString,com.alibaba.graphar.graphinfo.VertexInfo>)
during generating
        (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.alibaba.graphar.stdcxx.StdString,com.alibaba.graphar
.graphinfo.VertexInfo>) com.alibaba.graphar.graphinfo.GraphInfo::getVertexInfos()
        at com.alibaba.fastffi.annotation.TypeDefGenerator.reportError(TypeDefGenerator.java:212)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.reportError(TypeDefGenerator.java:208)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.reportError(TypeDefGenerator.java:204)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.getInternalTypeName(TypeDefGenerator.java:1451)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.generateFFIDelegateMethod(TypeDefGenerator.java:2302)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.generateFFIPointer(TypeDefGenerator.java:2004)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.generateFFIPointerOrFFILibrary(TypeDefGenerator.java:780)
        at com.alibaba.fastffi.annotation.TypeDefGenerator.generate(TypeDefGenerator.java:266)
        at com.alibaba.fastffi.annotation.TypeDefRegistry.lambda$generateFFIPointerFFILibrary$2(TypeDefRegistry.java:339)
        at com.alibaba.fastffi.annotation.TypeDefRegistry.processWorklist(TypeDefRegistry.java:325)
        at com.alibaba.fastffi.annotation.TypeDefRegistry.generateFFIPointerFFILibrary(TypeDefRegistry.java:337)
        at com.alibaba.fastffi.annotation.TypeDefRegistry.generate(TypeDefRegistry.java:302)
        at com.alibaba.fastffi.annotation.AnnotationProcessor.process(AnnotationProcessor.java:117)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:802)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:713)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1043)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1184)
        at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
        at com.sun.tools.javac.main.Main.compile(Main.java:523)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
        at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:126)
        at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:174)
        at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1129)
        at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:188)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /home/john/my-reop-gar/java/src/main/java/com/alibaba/graphar/graphinfo/GraphInfo.java:[197,33] Cannot find a TypeDef for
        std::map<std::string, GraphArchive::VertexInfo> -> (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.a
libaba.graphar.stdcxx.StdString,com.alibaba.graphar.graphinfo.VertexInfo>)
  during generating
        (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.alibaba.graphar.stdcxx.StdString,com.alibaba.graphar
.graphinfo.VertexInfo>) com.alibaba.graphar.graphinfo.GraphInfo::getVertexInfos()
  Type: com.alibaba.graphar.graphinfo.GraphInfo
  Executable: (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.alibaba.graphar.stdcxx.StdString,com.alibaba.g
raphar.graphinfo.VertexInfo>) com.alibaba.graphar.graphinfo.GraphInfo::getVertexInfos()
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.125 s
[ERROR]   during generating
[ERROR]         (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.alibaba.graphar.stdcxx.StdString,com.alibaba
.graphar.graphinfo.VertexInfo>) com.alibaba.graphar.graphinfo.GraphInfo::getVertexInfos()
[ERROR]   Type: com.alibaba.graphar.graphinfo.GraphInfo
[ERROR]   Executable: (@com.alibaba.fastffi.FFINameAlias("GetVertexInfos"),@com.alibaba.fastffi.CXXReference,@com.alibaba.fastffi.FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>") :: com.alibaba.graphar.stdcxx.StdMap<com.alibaba.graphar.stdcxx.StdString,com.a
libaba.graphar.graphinfo.VertexInfo>) com.alibaba.graphar.graphinfo.GraphInfo::getVertexInfos()
[ERROR]
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

File: java/src/main/java/com/alibaba/graphar/graphinfo/GraphInfo.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  /**
+   * Get the vertex infos of graph info
+   *
+   * @return vertex infos of graph info
+   */
+  //      @FFINameAlias("GetVertexInfos")
+  //      @FFIConst
+  //      @CXXReference
+  //      @FFITypeAlias("std::map<std::string, GraphArchive::VertexInfo>")
+  //      StdMap<StdString, VertexInfo> getVertexInfos();
```
Comment (Thespica) [inline java/src/main/java/com/alibaba/graphar/graphinfo/GraphInfo.java:196]: When compile with this function, met error : Cannot find a TypeDef for ....

Comment (Thespica) [reply java/src/main/java/com/alibaba/graphar/graphinfo/GraphInfo.java:196]: Thanks!
Both ways can solve this problem.

**Comment by Thespica** [comment]
PTAL :) @acezen @lixueclaire 

I think this PR is ready, but there are some problems when generating docs. In fact, I haven't changed existing docs neither add new docs.

It's maybe related to incompatibility between Python3 and jinja2. We need to discuss it in our weekly meeting.

**Comment by acezen** [comment]
hi, @Thespica, thanks for the job. Since you has added the unit test for java SDK, I think you should include the test to CI, so we can test the java code in every PR automatically.
The CI you can refer the C++ or Spark in here: https://github.com/alibaba/GraphAr/tree/main/.github/workflows

**Comment by lixueclaire** [comment]
@Thespica,  thank you for your valuable contribution! In addition to the tests, I believe it would enhance the user experience if a simple example is provided to showcase the usage of the Java SDK. Perhaps, this example could be included when you add the documentation.

**Comment by Thespica** [comment]
> hi, @Thespica, thanks for the job. Since you has added the unit test for java SDK, I think you should include the test to CI, so we can test the java code in every PR automatically. The CI you can refer the C++ or Spark in here: https://github.com/alibaba/GraphAr/tree/main/.github/workflows

Thanks for your advice! I well do it as soon as possible.

**Comment by Thespica** [comment]
> @Thespica, thank you for your valuable contribution! In addition to the tests, I believe it would enhance the user experience if a simple example is provided to showcase the usage of the Java SDK. Perhaps, this example could be included when you add the documentation.

Thanks for your advice too!

I think adding examples when adding documentation is better for me.

**Comment by acezen** [comment]
BTW，the known problem can be recorded in a issue that we can follow in the future.

**Review by acezen — CHANGES_REQUESTED** [review]
- The Java CI is a liitle complicated，simplify the CI with the builtin llvm and fastFFI. I have create a [java test CI](https://github.com/alibaba/GraphAr/blob/e7a4413c8a34f19e0f6fbbe58781e214d2d4b61d/.github/workflows/java.yml), You can refer to it。
- It's better to write a simple README to tell others that how to `BUILD`, `RUN` and `TEST` the java sdk, you can refer to the README of cpp

File: java/install-llvm11.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
Comment (acezen) [inline java/install-llvm11.sh:1]: Remove this file since we don't need to build llvm from the source

File: java/pom.xml [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        <dependency>
+            <groupId>com.alibaba.fastffi</groupId>
+            <artifactId>annotation-processor</artifactId>
+            <version>${fastffi.revision}</version>
+        </dependency>
+        <!-- The runtime component of LLVM4JNI -->
+        <dependency>
+            <groupId>com.alibaba.fastffi</groupId>
+            <artifactId>llvm4jni-runtime</artifactId>
+            <version>${fastffi.revision}</version>
```
Comment (acezen) [inline java/pom.xml:40]: add `<classifier>${os.detected.classifier}</classifier>` to install correct jar with os platform.

File: java/pom.xml [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            <artifactId>llvm4jni</artifactId>
+            <version>${fastffi.revision}</version>
+            <scope>provided</scope>
+        </dependency>
+
+
+        <dependency>
+            <groupId>com.alibaba.fastffi</groupId>
+            <artifactId>llvm</artifactId>
+            <version>${fastffi.revision}</version>
```
Comment (acezen) [inline java/pom.xml:54]: add <classifier>${os.detected.classifier}</classifier> to install correct jar with os platform.

File: java/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# GraphAr Java
+
```
Comment (acezen) [inline java/README.md:2]: Add a Working In Progress tag here to tell others that The JAVA SDK is working in progress and unfinished

**Review by acezen — APPROVED** [review]
LGTM~
You can merge after the `WIP` tag add to README

**Pull request merged**